### PR TITLE
Enable SNIP on multiple cards using DeepSpeed ZeRO-3

### PR DESCRIPTION
**Pull request by yiliu30**
## Type of Change

Feature and example.
API changed: None

## Description

The DeepSpeed Zero will [convert the module's parameters into ds_param](https://github.com/microsoft/DeepSpeed/blob/3e94f8c75116377d4b1c32b8c674368a27fb2a77/deepspeed/runtime/zero/partition_parameters.py#L1008) and override some parameter properties, such as `shape`, and `grad`. To `get` and `set` these properties, we need to use the [debug APIs](https://deepspeed.readthedocs.io/en/latest/zero3.html#debugging) provided by DeepSpeed.

This PR updates the `param.shape` and `param = new_tensor` with `safe_get_xx` and `safe_set_xx` accordingly, automatically calling the DeepSpeed debug APIs when using DeepSpeed.

## How has this PR been tested?

Pre-CI and local test.

## Dependency Change?
To run examples, DeepSpeed >= 0.12.4 is required.

### CODE REVIEW

File: neural_compressor/compression/pruner/criteria.py [hunk 1]
```diff
@@ -96,7 +99,9 @@ def on_step_begin(self):
         """Calculate and store the pruning scores based on a magnitude criterion."""
         with torch.no_grad():
             for key in self.modules.keys():
-                p = self.modules[key].weight.data
+                param = self.modules[key].weight
+                data = safe_get_data(param)
+                p = data
```
Comment (WeiweiZhang1) [inline neural_compressor/compression/pruner/criteria.py:104]: p =  safe_get_data(param)

File: neural_compressor/compression/pruner/patterns/base.py [hunk 2]
```diff
@@ -20,6 +20,8 @@
 
 import numpy as np
 
+from neural_compressor.compression.pruner.utils import safe_get_data, safe_get_grad, safe_get_shape
```
Comment (WeiweiZhang1) [inline neural_compressor/compression/pruner/patterns/base.py:23]: for better compatibility of the pruning module on ITREX, it's preferable to replace the import path containing 'neural_compressor'
such as:
from ..utils import safe_get_data, safe_get_grad, safe_get_shape

File: neural_compressor/compression/pruner/patterns/ninm.py [hunk 3]
```diff
@@ -145,12 +147,22 @@ def _reshape_orig_to_2dims(self, data):
         Returns:
             Reshaped data.
         """
-        if len(data.shape) == 4:  # TODO: need to verify whether it's ok for transposed conv
+        from neural_compressor.compression.pruner.utils import FLATTEN_DIM2
```
Comment (WeiweiZhang1) [inline neural_compressor/compression/pruner/patterns/ninm.py:150]: from ..utils import FLATTEN_DIM2

**Review by WeiweiZhang1 — APPROVED** [review]

File: neural_compressor/compression/pruner/patterns/base.py (hunk 2 above)
Comment (yiliu30) [reply neural_compressor/compression/pruner/patterns/base.py]: nice catch, updated in [ed6395b](https://github.com/intel/neural-compressor/pull/1492/commits/ed6395b2fa9ae9aecce68b2a0ceb6e3e1ec2f02d).

File: neural_compressor/compression/pruner/patterns/ninm.py (hunk 3 above)
Comment (yiliu30) [reply neural_compressor/compression/pruner/patterns/ninm.py]: nice catch, updated in [ed6395b](https://github.com/intel/neural-compressor/pull/1492/commits/ed6395b2fa9ae9aecce68b2a0ceb6e3e1ec2f02d).

File: neural_compressor/compression/pruner/criteria.py (hunk 1 above)
Comment (yiliu30) [reply neural_compressor/compression/pruner/criteria.py]: Thanks, updated in [ed6395b](https://github.com/intel/neural-compressor/pull/1492/commits/ed6395b2fa9ae9aecce68b2a0ceb6e3e1ec2f02d).

**Review by lkk12014402 — APPROVED** [review]

**Pull request merged**